File: cpp_files/raylib/shape_survivors/src/bullets/bullet_manager.cpp
#include "bullet_manager.hpp"

static const int BULLETS_MAX = 200;

BulletManager::BulletManager() : spawner({.max = BULLETS_MAX}) {}

void BulletManager::spawn(const BulletConfig& config,
                          const Vector2&      start_pos,
                          const Vector2&      target_pos) {
    const int bullets_size = static_cast<int>(bullets.size());

    if (bullets_size >= BULLETS_MAX)
        return;

    bullets.push_back(Bullet(config, start_pos, target_pos));
}

void BulletManager::despawn(const int index) {
    const int bullets_size = static_cast<int>(bullets.size());

    if (index < 0 || index >= bullets_size)
        return;

    bullets[index] = bullets[bullets_size - 1];
    bullets.pop_back();
}

void BulletManager::update(const float dt) {
    const int bullets_size = static_cast<int>(bullets.size());
    for (int i = 0; i < bullets_size; i++) {
        Bullet& bullet = bullets[i];

        bullet.update(dt);

        if (bullet.lifetime.current <= 0) {
            despawn(i);
            i--;
        }
    }
}

void BulletManager::draw(void) const {
    for (const Bullet& bullet : bullets)
        bullet.draw();
}

File: python_files/raylib_files/shape_survivors/bullet_manager.py
import pyray as p

from bullet import Bullet, BulletConfig

BULLETS_MAX: int = 200


class BulletManager:
    def __init__(self) -> None:
        self.bullets: list[Bullet] = []

    def spawn(
        self, config: BulletConfig, start_pos: p.Vector2, target_pos: p.Vector2
    ) -> None:
        if len(self.bullets) >= BULLETS_MAX:
            return

        self.bullets.append(Bullet(config, start_pos, target_pos))

    def despawn(self, index: int) -> None:
        if index < 0 or index >= len(self.bullets):
            return

        self.bullets[index] = self.bullets[len(self.bullets) - 1]
        self.bullets.pop()

    def update(self, dt: float) -> None:
        bullet_removal_list: list[int] = []
        for index, bullet in enumerate(self.bullets):
            bullet.update(dt)

            if bullet.lifetime.current <= 0:
                bullet_removal_list.append(index)

        for index in bullet_removal_list:
            self.despawn(index)

    def draw(self) -> None:
        for bullet in self.bullets:
            bullet.draw()
